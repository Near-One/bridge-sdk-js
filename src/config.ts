export type NetworkType = "mainnet" | "testnet"

const ADDRESSES = {
  mainnet: {
    arb: "0xd025b38762B4A4E36F0Cde483b86CB13ea00D989",
    base: "0xd025b38762B4A4E36F0Cde483b86CB13ea00D989",
    bnb: "0x073C8a225c8Cf9d3f9157F5C1a1DbE02407f5720",
    eth: "0xe00c629afaccb0510995a2b95560e446a24c85b9",
    near: "omni.bridge.near",
    sol: {
      locker: "dahPEoZGXfyV58JqqH85okdHmpN8U2q8owgPUXSCPxe",
      wormhole: "worm2ZoG2kUd4vFXhvjh93UUH596ayRfgQ2MgjNMTth",
      shimProgram: "EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX",
      eventAuthority: "HQS31aApX3DDkuXgSpV9XyDUNtFgQ31pUn5BNWHG2PSp",
    },
    btc: {
      network: "mainnet" as const,
      apiUrl: "https://blockstream.info/api",
      mempoolUrl: "https://mempool.space/api",
      rpcUrl: "https://bitcoin-rpc.publicnode.com",
      btcConnector: "btc-connector.bridge.near",
      btcToken: "nbtc.bridge.near",
      bitcoinRelayer: "satoshi_optwo.near",
    },
    zcash: {
      network: "mainnet" as const,
      apiUrl: "https://zcash-mainnet.gateway.tatum.io/",
      rpcUrl: "https://zcash-mainnet.gateway.tatum.io/",
      zcashConnector: "zcash-connector.bridge.near",
      zcashToken: "nzec.bridge.near",
    },
  },
  testnet: {
    arb: "0x0C981337fFe39a555d3A40dbb32f21aD0eF33FFA",
    base: "0xa56b860017152cD296ad723E8409Abd6e5D86d4d",
    bnb: "0x7Fd1E9F9ed48ebb64476ba9E06e5F1a90e31DA74",
    eth: "0x68a86e0Ea5B1d39F385c1326e4d493526dFe4401",
    near: "omni.n-bridge.testnet",
    sol: {
      locker: "862HdJV59Vp83PbcubUnvuXc4EAXP8CDDs6LTxFpunTe",
      wormhole: "3u8hJUVTA4jH1wYAyUur7FFZVQ8H635K3tSHHF4ssjQ5",
      shimProgram: "EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX",
      eventAuthority: "HQS31aApX3DDkuXgSpV9XyDUNtFgQ31pUn5BNWHG2PSp",
    },
    btc: {
      network: "testnet" as const,
      apiUrl: "https://blockstream.info/testnet/api",
      mempoolUrl: "https://mempool.space/testnet/api",
      rpcUrl: "https://bitcoin-testnet-rpc.publicnode.com",
      btcConnector: "brg-dev.testnet",
      btcToken: "nbtc.n-bridge.testnet",
      bitcoinRelayer: "cosmosfirst.testnet",
    },
    zcash: {
      network: "testnet" as const,
      apiUrl: "https://zcash-testnet.gateway.tatum.io/",
      rpcUrl: "https://zcash-testnet.gateway.tatum.io/",
      zcashConnector: "zcash_connector.n-bridge.testnet",
      zcashToken: "nzcash.n-bridge.testnet",
    },
  },
} as const

let selectedNetwork: NetworkType = "mainnet"

export function setNetwork(network: NetworkType) {
  selectedNetwork = network
}

export function getNetwork(): NetworkType {
  return selectedNetwork
}

export const addresses = {
  get arb() {
    return ADDRESSES[selectedNetwork].arb
  },
  get base() {
    return ADDRESSES[selectedNetwork].base
  },
  get bnb() {
    return ADDRESSES[selectedNetwork].bnb
  },
  get eth() {
    return ADDRESSES[selectedNetwork].eth
  },
  get near() {
    return ADDRESSES[selectedNetwork].near
  },
  get sol() {
    return ADDRESSES[selectedNetwork].sol
  },
  get btc() {
    return ADDRESSES[selectedNetwork].btc
  },
  get zcash() {
    return ADDRESSES[selectedNetwork].zcash
  },
  get network() {
    return selectedNetwork
  },
}
