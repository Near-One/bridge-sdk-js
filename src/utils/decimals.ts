import { addresses } from "../config"
import type { OmniAddress } from "../types"

export interface TokenDecimals {
  decimals: number
  origin_decimals: number
}

/**
 * Normalizes an amount from one decimal precision to another using BigInt math
 * @param amount The amount to normalize as a bigint
 * @param fromDecimals The source decimal precision
 * @param toDecimals The target decimal precision
 * @returns The normalized amount as a bigint
 */
export function normalizeAmount(amount: bigint, fromDecimals: number, toDecimals: number): bigint {
  if (fromDecimals === toDecimals) return amount

  if (fromDecimals > toDecimals) {
    // Scale down: Divide by power of 10
    const scale = 10n ** BigInt(fromDecimals - toDecimals)
    return amount / scale
    // biome-ignore lint/style/noUselessElse: Adds clarity
  } else {
    // Scale up: Multiply by power of 10
    const scale = 10n ** BigInt(toDecimals - fromDecimals)
    return amount * scale
  }
}

/**
 * Verifies if a transfer amount will be valid after normalization
 * @param amount The amount to transfer
 * @param fee The fee to be deducted
 * @param originDecimals The decimals of the token on the source chain
 * @param destinationDecimals The decimals of the token on the destination chain
 * @returns true if the normalized amount (minus fee) will be greater than 0
 */
export function verifyTransferAmount(
  amount: bigint,
  fee: bigint,
  originDecimals: number,
  destinationDecimals: number,
): boolean {
  try {
    // Use the minimum of origin and destination decimals for normalization
    const minDecimals = Math.min(originDecimals, destinationDecimals)

    // First normalize amount and fee to the minimum decimals
    const normalizedAmount = normalizeAmount(amount, originDecimals, minDecimals)
    const normalizedFee = normalizeAmount(fee, originDecimals, minDecimals)

    // Check if amount minus fee is greater than 0
    return normalizedAmount > normalizedFee && normalizedAmount - normalizedFee > 0n
  } catch {
    // If we hit any math errors, the amount is effectively too small
    return false
  }
}

/**
 * Get the minimum number of decimals between origin and destination chains
 * @param contractId The NEAR contract to query
 * @param sourceToken The source token address
 * @param destinationToken The destination token address
 * @returns The minimum number of decimals
 */
export async function getMinimumDecimals(
  contractId: string,
  sourceToken: OmniAddress,
  destinationToken: OmniAddress,
): Promise<number> {
  const sourceDecimals = await getTokenDecimals(contractId, sourceToken)
  const destDecimals = await getTokenDecimals(contractId, destinationToken)

  return Math.min(sourceDecimals.decimals, destDecimals.decimals)
}

/**
 * Gets token decimals from the NEAR contract
 * @param contractId The NEAR contract ID to query
 * @param tokenAddress The Omni token address to check
 * @returns Promise resolving to the token's decimal information
 */
export async function getTokenDecimals(
  contractId: string,
  tokenAddress: OmniAddress,
): Promise<TokenDecimals> {
  const response = await fetch(`https://rpc.${addresses.network}.near.org`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      jsonrpc: "2.0",
      id: "dontcare",
      method: "query",
      params: {
        request_type: "call_function",
        finality: "final",
        account_id: contractId,
        method_name: "get_token_decimals",
        args_base64: Buffer.from(JSON.stringify({ address: tokenAddress })).toString("base64"),
      },
    }),
  })

  const result = await response.json()

  if (result.error) {
    throw new Error(
      `Failed to get token decimals: ${result.error.message || JSON.stringify(result.error)}`,
    )
  }

  // Parse the result which is returned as an encoded JSON string
  const decoded = Buffer.from(result.result.result).toString()
  const decimals = JSON.parse(decoded) as TokenDecimals

  return decimals
}

/**
 * Gets the minimum transferable amount for a token pair accounting for decimal normalization
 * @param originDecimals The decimals of the source token
 * @param destinationDecimals The decimals of the destination token
 * @returns The minimum transferable amount as a bigint
 */
export function getMinimumTransferableAmount(
  originDecimals: number,
  destinationDecimals: number,
): bigint {
  // Start with 1 in destination decimal system
  let minAmount = 1n

  // If origin has more decimals, we need to scale up
  if (originDecimals > destinationDecimals) {
    minAmount = minAmount * 10n ** BigInt(originDecimals - destinationDecimals)
  }

  return minAmount
}
