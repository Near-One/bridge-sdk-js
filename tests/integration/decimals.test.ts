import { afterAll, beforeAll, describe, it, vi } from "vitest"
import { setNetwork } from "../../src/config.js"
import { getTokenDecimals } from "../../src/utils/decimals.js"

describe.concurrent("getTokenDecimals integration", () => {
  setNetwork("testnet")

  beforeAll(() => {
    vi.spyOn(console, "error").mockImplementation(() => {})
    vi.spyOn(console, "debug").mockImplementation(() => {})
  })
  afterAll(() => {
    vi.restoreAllMocks()
  })

  it("fetches decimals for JLU token on Solana", async ({ expect }) => {
    const tokenAddress = "sol:rLFLkpdMZsVLWziDfz5WWqVgVnFbPdKicSNQcj9QVxL"
    const decimals = await getTokenDecimals("omni-locker.testnet", tokenAddress)

    // Verify response structure (these should not change)
    expect(decimals).not.toBeNull()
    if (decimals) {
      expect(decimals).toHaveProperty("decimals")
      expect(decimals).toHaveProperty("origin_decimals")
      expect(typeof decimals.decimals).toBe("number")
      expect(typeof decimals.origin_decimals).toBe("number")

      // Snapshot the actual values
      expect(decimals).toMatchSnapshot({
        decimals: 9,
        origin_decimals: 18,
      })
    }
  }, 10000) // Increase timeout for RPC call

  it("handles invalid token addresses", async ({ expect }) => {
    const invalidAddress = "sol:invalid.testnet"
    await expect(getTokenDecimals("omni-locker.testnet", invalidAddress)).rejects.toMatchSnapshot()
  })

  it("returns null for unregistered token addresses", async ({ expect }) => {
    const unregisteredAddress = "sol:unregistered1234567890"
    const result = await getTokenDecimals("omni-locker.testnet", unregisteredAddress)
    // This might return null for unregistered tokens
    if (result === null) {
      expect(result).toBeNull()
    } else {
      // If it doesn't return null, it should be a valid TokenDecimals object
      expect(result).toHaveProperty("decimals")
      expect(result).toHaveProperty("origin_decimals")
    }
  })
})
