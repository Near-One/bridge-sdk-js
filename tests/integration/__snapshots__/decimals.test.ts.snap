// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTokenDecimals integration > fetches decimals for JLU token on Solana 1`] = `
{
  "decimals": 9,
  "origin_decimals": 18,
}
`;

exports[`getTokenDecimals integration > handles invalid token addresses 1`] = `[TypeError: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received undefined]`;
